File: easynav_workshop/easynav_patrolling_behavior_py/easynav_patrolling_behavior_py/patrolling_node.py
# Copyright 2025 Intelligent Robotics Lab
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from enum import Enum

import math
from typing import List, Tuple, Optional

import rclpy
try:
    from easynav_goalmanager_py import GoalManagerClient, ClientState  # type: ignore
except Exception:  # pragma: no cover
    GoalManagerClient = None
    ClientState = None

from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Goals
from tf_transformations import quaternion_from_euler

class PatrolState(Enum):
    IDLE = 0
    PATROLLING = 1
    DO_AT_WAYPOINT = 2
    FINISHED = 3
    ERROR = 4

def quat_from_yaw(yaw: float) -> Tuple[float, float, float, float]:
    cy = math.cos(yaw * 0.5)
    sy = math.sin(yaw * 0.5)
    return (0.0, 0.0, sy, cy)

class PatrollingNode(Node):
    def __init__(self):
        super().__init__('patrolling_node')

        wp = ['']

        self.declare_parameter('frame_id', 'map')
        self.declare_parameter('waypoints', wp)

        self.frame_id: str = self.get_parameter('frame_id').get_parameter_value().string_value

        wp_names = self.get_parameter('waypoints').get_parameter_value().string_array_value

        self._goals = Goals()
        self._goals.header.frame_id = self.frame_id
        self._goals.header.stamp = self.get_clock().now().to_msg()
        
        for name in wp_names:
            vals = [0.0, 0.0, 0.0]
            self.declare_parameter(name, vals)
            param = self.get_parameter(name)
            if not param:
                self.get_logger().warn(f'Missing waypoint parameter: {name}')
                continue
            vals = param.get_parameter_value().double_array_value
            if len(vals) != 3:
                self.get_logger().warn(f'Waypoint {name} must have 3 values [x, y, yaw]')
                continue

            new_wp = PoseStamped()
            new_wp.header.frame_id = self.frame_id
            new_wp.pose.position.x = vals[0]
            new_wp.pose.position.y = vals[1]

            q = quaternion_from_euler(0, 0, vals[2])

            new_wp.pose.orientation.x = q[0]
            new_wp.pose.orientation.y = q[1]
            new_wp.pose.orientation.z = q[2]
            new_wp.pose.orientation.w = q[3]

            self._goals.goals.append(new_wp)

        self._gm = GoalManagerClient(node=self)
        self.current_goal_index = 0
        self._timer = self.create_timer(0.2, self._cycle)
        self.get_logger().info('PatrollingNode started')
        self._state = PatrolState.IDLE



    def _cycle(self):
        match self._state:
            case PatrolState.IDLE:
                single_goal = Goals()
                single_goal.goals.append(self._goals.goals[self.current_goal_index])
                single_goal.header = self._goals.header
                self.get_logger().info(f'Sending goal to waypoint {self.current_goal_index + 1}')
                self._gm.send_goals(single_goal)
                self._state = PatrolState.PATROLLING

            case PatrolState.PATROLLING:
                nav_state = self._gm.get_state()
                if (
                    nav_state == ClientState.NAVIGATION_REJECTED or
                    nav_state == ClientState.NAVIGATION_FAILED or
                    nav_state == ClientState.NAVIGATION_CANCELLED or
                    nav_state == ClientState.ERROR
                ):
                    result = self._gm.get_result()
                    self.get_logger().debug(
                        f'Navigation finished with error {result.status_message}')
                    self._state = PatrolState.ERROR
                elif nav_state == ClientState.NAVIGATION_FINISHED:
                    result = self._gm.get_result()
                    self.get_logger().info(
                        f'Navigation succesfully finished with message {result.status_message}')
                    
                    self.pause_start_time = self.get_clock().now()
                   
                    self.get_logger().info(f"Waiting time started at waypoint {self.current_goal_index + 1}")
                    self._state = PatrolState.DO_AT_WAYPOINT

                elif nav_state == ClientState.ACCEPTED_AND_NAVIGATING:
                    pass

                else:
                    pass

            case PatrolState.DO_AT_WAYPOINT:
                if self.get_clock().now() - self.pause_start_time >= self.pause_duration:
                    self.get_logger().info(f"Waiting time ended at waypoint {self.current_goal_index + 1}")

                    # Advance to the next waypoint
                    self.current_goal_index += 1
                    if self.current_goal_index < len(self._goals.goals):
                        self._gm.reset()
                        self._state = PatrolState.IDLE
                    else:
                        self.get_logger().info("All waypoints completed")
                        self._state = PatrolState.FINISHED

            case PatrolState.FINISHED:
                self.get_logger().info('Reset navigation')
                self.current_goal_index = 0
                nav_state = self._gm.get_state()
                if nav_state != ClientState.IDLE:
                    self._gm.reset()
                self._state = PatrolState.IDLE
            case PatrolState.ERROR:
                pass


def main(args=None):
    rclpy.init(args=args)
    node = PatrollingNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: exercises/easynav/easynav_playground/easynav_patrolling_behavior/src/easynav_patrolling_behavior/PatrollingNode.cpp
// Copyright 2025 Intelligent Robotics Lab
//
// This file is part of the project Easy Navigation (EasyNav in short)
// licensed under the GNU General Public License v3.0.
// See <http://www.gnu.org/licenses/> for details.
//
// Easy Navigation program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/macros.hpp"

#include "geometry_msgs/msg/pose_stamped.hpp"

#include "easynav_system/GoalManagerClient.hpp"
#include "easynav_patrolling_behavior/PatrollingNode.hpp"

namespace easynav
{

    using namespace std::chrono_literals;

    PatrollingNode::PatrollingNode(const rclcpp::NodeOptions &options)
        : Node("patrolling_node", options)
    {
        timer_ = create_timer(
            100ms,
            std::bind(&PatrollingNode::cycle, this));
    }

    void
    PatrollingNode::initialize()
    {
        std::vector<std::string> waypoints;
        declare_parameter("waypoints", waypoints);
        get_parameter("waypoints", waypoints);

        declare_parameter<std::string>("frame_id", "map");
        get_parameter("frame_id", frame_id_);

        goals_.header.frame_id = frame_id_;
        
        for (const auto &wp : waypoints)
        {
            std::vector<double> wp_coord;
            declare_parameter(wp, wp_coord);
            get_parameter(wp, wp_coord);

            if (wp_coord.size() != 3)
            {
                RCLCPP_ERROR(get_logger(), "Coordinates for wp [%s] have wrong size %zu",
                             wp.c_str(), wp_coord.size());
                continue;
            }

            geometry_msgs::msg::PoseStamped wp_pose;
            wp_pose.header.frame_id = frame_id_;
            wp_pose.pose.position.x = wp_coord[0];
            wp_pose.pose.position.y = wp_coord[1];
            wp_pose.pose.orientation = orientationAroundZAxis(wp_coord[2]);

            goals_.goals.push_back(wp_pose);
        }
    }
    nav_msgs::msg::Goals
    PatrollingNode::build_current_goal()
    {
        nav_msgs::msg::Goals single_goal;
        single_goal.header = goals_.header;
        single_goal.goals.push_back(goals_.goals[current_goal_index_]);
        return single_goal;
    }

    void
    PatrollingNode::cycle()
    {
        switch (state_)
        {
        case PatrolState::IDLE:
        {
            if (!initialized_)
            {
                RCLCPP_INFO(get_logger(), "Initializing patrolling");
                gm_client_ = GoalManagerClient::make_shared(shared_from_this());
                initialize();
                initialized_ = true;
            }
            RCLCPP_INFO(get_logger(), "Sending goals to waypoint %zu", current_goal_index_ + 1);
            gm_client_->send_goals(build_current_goal());
            state_ = PatrolState::PATROLLING;
        }
        break;
        
        case PatrolState::PATROLLING:
        {
            auto nav_state = gm_client_->get_state();
            switch (nav_state)
            {
            case GoalManagerClient::State::NAVIGATION_REJECTED:
            case GoalManagerClient::State::NAVIGATION_FAILED:
            case GoalManagerClient::State::NAVIGATION_CANCELLED:
            case GoalManagerClient::State::ERROR:
                RCLCPP_ERROR(get_logger(), "Navigation finished with error %s",
                             gm_client_->get_result().status_message.c_str());
                state_ = PatrolState::ERROR;
                break;

            case GoalManagerClient::State::SENT_GOAL:
                if (send_retries_ < max_retries_)
                {
                    send_retries_++; // Waiting for ACCEPT
                }
                else
                {
                    send_retries_ = 0;
                    state_ = PatrolState::IDLE; // No accept, retry sending goal
                }
                break;

            case GoalManagerClient::State::NAVIGATION_FINISHED:
                RCLCPP_INFO(get_logger(), "Navigation succesfully finished with message %s",
                            gm_client_->get_result().status_message.c_str());
                
                state_ = PatrolState::DO_AT_WAYPOINT;
                break;
            case GoalManagerClient::State::ACCEPTED_AND_NAVIGATING:
                break;
            default:
                break;
            }
        }
        break;

        case PatrolState::DO_AT_WAYPOINT:
            /// TODO:
            // Implement actions at waypoint before proceeding to the next one
            // You could add a wait time emulating perform specific tasks here
            // You could log data, spin the robot in place, etc.
            // Remember to transition to the next IDLE state after completing intermediate actions
            // or to the finished state if all waypoints have been visited
            // and increment the current_goal_index_ accordingly
            break;

        case PatrolState::FINISHED:
            RCLCPP_INFO(get_logger(), "Reset navigation");
            current_goal_index_ = 0;
            state_ = PatrolState::IDLE;
            if (gm_client_->get_state() != GoalManagerClient::State::IDLE)
            {
                gm_client_->reset();
            }
            break;
        case PatrolState::ERROR:
            break;
        }
    }

} // namespace easynav
